File: LeetCode/LeetCode Practice Problems/1008- Construct Binary Search Tree from Preorder Traversal.cpp
class Solution {
public:
    TreeNode* head;
    TreeNode* insert(TreeNode* h,int v)
    {
        if(h == NULL)
        {
            TreeNode* r = new TreeNode(v);
            return r;
        }
        
        if(h->val < v)
            h->right = insert(h->right , v);
        else
            h->left = insert(h->left , v);
        
        return h;
    }
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        for(auto i : preorder)
            head = insert(head , i);
        
        return head;
    }
};

File: LeetCode/LeetCode Practice Problems/1008-Construct Binary Search Tree from Preorder Traversal.java
class Solution {
    
    public TreeNode constructTree(int key,TreeNode root) {
        if(root == null) {
            return new TreeNode(key);
        }
        
        if(root.val > key) {
            root.left = constructTree(key , root.left);
        } else {
            root.right = constructTree(key , root.right);
        }
        
        return root;
    }
    
    public TreeNode bstFromPreorder(int[] preorder) {
        TreeNode root = null;
        
        for(int key: preorder) {
            root = constructTree(key , root);
        }
        
        return root;
    }
}
